src/compiler/checker.ts

```diff
@@ -28417,6 +28417,47 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkObjectLiteral(node: ObjectLiteralExpression, checkMode?: CheckMode): Type {
+        const contextualType = getContextualType(node, /*contextFlags*/ undefined);
+        const isContextualTypeDependent = 
+            contextualType &&
+            contextualType.immediateBaseConstraint && 
+            contextualType.immediateBaseConstraint.aliasTypeArguments &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments.length === 1 &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments[0].id === contextualType.id;
+
+        if (!isContextualTypeDependent) {
+            return checkObjectLiteralNonDependently(node, checkMode);
+        }
+
+        let valueType = checkObjectLiteralNonDependently(node, checkMode);
+
+        let newContextualType = cloneTypeParameter(contextualType);
+        newContextualType.immediateBaseConstraint =
+            instantiateType(
+                contextualType.immediateBaseConstraint,
+                createTypeMapper([contextualType], [valueType])
+            );

```

Comment:
This doesn't seem to replace the types from default parameters in a type alias, which is not what I would have expected. That is extracting out `Exclude<keyof Self, "k" | "t">` to a default type parameter `K` (as I did in #40439) makes the test fail even though it doesn't change anything.